### PR TITLE
Specify which Inline query failed

### PR DESCRIPTION
**Pull request by saolsen**
PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/README.md [hunk 1]
```diff
@@ -83,6 +83,7 @@ Execute a Polar query through the FFI event interface.
 | `results`           |        |         | x            |         |
 | `getCall`           |        |         | x            |         |
 | `run`               | `run`  | `start` | `nextResult` | `start` |
+| `source`            | x      | x       | x            |         |
```
Comment (saolsen) [inline languages/README.md:86]:
```suggestion
| `source`            | x      | x       | x            | x       |
```

File: polar-core/src/polar.rs [hunk 2]
```diff
@@ -36,6 +37,10 @@ impl Query {
     pub fn next_message(&self) -> Option<Message> {
         self.vm.messages.next()
     }
+
+    pub fn source(&self) -> String {
+        self.vm.term_source(&self.source)
```
Comment (samscott89) [inline polar-core/src/polar.rs:42]: Maybe we should include the context too? From `source.source`? Will have line number for example.

**Comment by saolsen** [comment]
Messages look like this now. (example in python)
`polar.exceptions.PolarRuntimeException: Inline query in file failed. 1 = 2 at line 1, column 3 in file foo.polar`

File: languages/js/src/Query.ts [hunk 3]
```diff
@@ -38,6 +38,10 @@ export class Query {
     this.results = this.start();
   }
 
+  public source(): string {
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/js/src/Query.ts:41]: Also please add API docs -- see the rest of the methods in this file for examples. This one should probably be marked `@internal`

File: languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java [hunk 4]
```diff
@@ -225,8 +225,8 @@ public InstantiationError(String className, Exception e) {
     }
 
     public static class InlineQueryFailedError extends PolarRuntimeException {
-        public InlineQueryFailedError() {
-            super(null, null);
+        public InlineQueryFailedError(String source) {
+            super("Inline Query Failed " + source);
```
Comment (gj) [inline languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java:229]: I think it would be worth standardizing this error messages across the libs. I know we need a general standardization pass that goes beyond this single error type, but no reason not to go for incremental improvement.

File: polar-core/src/vm.rs [hunk 5]
```diff
@@ -694,7 +694,7 @@ impl PolarVirtualMachine {
         }
     }
 
-    fn source(&self, term: &Term) -> Option<Source> {
+    pub fn source(&self, term: &Term) -> Option<Source> {
```
Comment (gj) [inline polar-core/src/vm.rs:697]: Why does this have to be `pub`?

File: languages/js/src/Polar.ts [hunk 6]
```diff
@@ -151,10 +151,13 @@ export class Polar {
...
       this.processMessages();
       if (query === undefined) break;
+      const source = query.source();
       const { results } = new Query(query, this.#host);
       const { done } = await results.next();
       results.return();
-      if (done) throw new InlineQueryFailedError(name);
+      if (done) {
+        throw new InlineQueryFailedError(source, name);
+      }
```
Comment (gj) [inline languages/js/src/Polar.ts:160]: This is the style we use across the rest of the lib
```suggestion
      if (done) throw new InlineQueryFailedError(source, name);
```

File: languages/js/src/Query.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gj) [reply languages/js/src/Query.ts:41]: Now that I look at it more closely, we aren't actually using this method at all. You're calling `.source()` directly on the FFI query instead of on the TS `Query` instance.

File: languages/js/src/errors.ts [hunk 7]
```diff
@@ -35,11 +35,11 @@ export class DuplicateInstanceRegistrationError extends PolarError {
...
 
 export class InlineQueryFailedError extends PolarError {
-  constructor(file?: string) {
+  constructor(source: string, file?: string) {
     if (file !== undefined) {
-      super(`Inline query failed while loading '${file}'.`);
+      super(`Inline query failed while loading '${file}', ${source}.`);
     } else {
-      super('Inline query failed.');
+      super(`Inline query failed, ${source}.`);
```
Comment (gj) [inline languages/js/src/errors.ts:42]: I think these commas would be better as colons, maybe with the source indented on the following line. Something like:
```
Inline query failed while loading file policies/production.polar:
  ?= 1 = 2;
```

File: languages/ruby/lib/oso/polar/ffi/query.rb [hunk 8]
```diff
@@ -74,6 +75,15 @@ def process_messages
             message.process
           end
         end
+
+        # @return [String]
```
Comment (gj) [inline languages/ruby/lib/oso/polar/ffi/query.rb:79]: This doesn't look like it returns a string -- did you mean to return `res.to_s`?

File: languages/ruby/lib/oso/polar/ffi/query.rb [hunk 9]
```diff
@@ -74,6 +75,15 @@ def process_messages
...
+        # @raise [FFI::Error] if the FFI call returns an error.
+        def source
+          res = Rust.source(self)
+          raise FFI::Error.get if res.null?
```
Comment (gj) [inline languages/ruby/lib/oso/polar/ffi/query.rb:83]: Judging by your CFFI code in polar-c-api/src/lib.rs, it doesn't look like it's possible for this FFI call to return null

File: polar-core/src/polar.rs [hunk 10]
```diff
@@ -36,6 +37,19 @@ impl Query {
     pub fn next_message(&self) -> Option<Message> {
         self.vm.messages.next()
     }
+
+    pub fn source(&self) -> String {
+        let mut source_str = self.vm.term_source(&self.source);
+        if let Some(source) = self.vm.source(&self.source) {
```
Comment (gj) [inline polar-core/src/polar.rs:43]: Could you explain what's going on here? It's a bit hard to follow - the `Query#source` method gets the `term_source` for its own `source` field which calls `VM#source` to get the term source info and then we call `VM#source` again (???), passing in the query's `source` field again... and I've already lost the plot of what this is actually supposed to result in 😄

File: polar-core/src/vm.rs (hunk 5 above)
Comment (saolsen) [reply polar-core/src/vm.rs:697]: So we can call it from the method on Query.

File: languages/ruby/lib/oso/polar/ffi/query.rb (hunk 9 above)
Comment (saolsen) [reply languages/ruby/lib/oso/polar/ffi/query.rb:83]: If there's a panic it would still return null.

File: polar-core/src/polar.rs (hunk 10 above)
Comment (saolsen) [reply polar-core/src/polar.rs]: Renamed a bunch of stuff.

File: languages/README.md [hunk 11]
```diff
@@ -83,6 +83,7 @@ Execute a Polar query through the FFI event interface.
 | `results`           |        |         | x            |         |
 | `getCall`           |        |         | x            |         |
 | `run`               | `run`  | `start` | `nextResult` | `start` |
+| `source`            | x      | x       | x            | x       |
```
Comment (gj) [inline languages/README.md:86]: This isn't a method on the Query object -- we're making a direct FFI call.

File: languages/js/src/errors.ts [hunk 12]
```diff
@@ -35,12 +35,8 @@ export class DuplicateInstanceRegistrationError extends PolarError {
...
 
 export class InlineQueryFailedError extends PolarError {
-  constructor(file?: string) {
-    if (file !== undefined) {
-      super(`Inline query failed while loading '${file}'.`);
-    } else {
-      super('Inline query failed.');
-    }
+  constructor(source: string) {
+    super(`Inline query failed: ${source}`);
```
Comment (gj) [inline languages/js/src/errors.ts:39]: Did we remove the optional filename handling because `source` will contain the same info?

File: languages/python/oso/polar/polar.py [hunk 13]
```diff
@@ -140,7 +140,8 @@ def load_str(self, string, filename=None):
                 try:
                     next(Query(query, host=self.host.copy()).run())
                 except StopIteration:
-                    raise PolarRuntimeException("Inline query in file failed.")
+                    source = query.source()
+                    raise PolarRuntimeException(f"Inline query failed: {source.get()}")
```
Comment (gj) [inline languages/python/oso/polar/polar.py:144]: Nit: it would be good to create an error type for this so that people can rescue it specifically without matching on the message, but that can be part of a (necessary) broader Python error handling cleanup.

File: languages/ruby/lib/oso/polar/polar.rb [hunk 14]
```diff
@@ -105,7 +105,7 @@ def load_str(str, filename: nil) # rubocop:disable Metrics/MethodLength
           begin
             Query.new(next_query, host: host).results.next
           rescue StopIteration
-            raise InlineQueryFailedError
+            raise InlineQueryFailedError, format('Inline query failed: %<source>s', source: next_query.source)
```
Comment (gj) [inline languages/ruby/lib/oso/polar/polar.rb:108]: This formatting should be handled in the `InlineQueryFailedError#initialize` method:
```rb
    class InlineQueryFailedError < PolarRuntimeError # rubocop:disable Style/Documentation
      # @param file [String]
      def initialize(source)
        super(format('Inline query failed: %<source>s', source: source))
      end
    end
```

and then this becomes:
```suggestion
            raise InlineQueryFailedError, next_query.source
```

File: polar-core/src/polar.rs [hunk 15]
```diff
@@ -36,6 +37,19 @@ impl Query {
     pub fn next_message(&self) -> Option<Message> {
         self.vm.messages.next()
     }
+
+    pub fn source_info(&self) -> String {
+        let mut source_info = self.vm.term_source_string(&self.term);
+        if let Some(source) = self.vm.source(&self.term) {
```
Comment (gj) [inline polar-core/src/polar.rs:43]: It still seems suboptimal that we're calling `self.vm.term_source_string(&self.term)`, which calls `self.vm.source(&self.term)`, and then we're calling `self.vm.source(&self.term)` _again_

File: languages/js/src/errors.ts (hunk 12 above)
Comment (saolsen) [reply languages/js/src/errors.ts:39]: Yeah. If it's from a file the location and filename are part of the message.

File: languages/python/oso/polar/exceptions.py [hunk 16]
```diff
@@ -45,6 +45,10 @@ class PolarFileNameChangedError(PolarRuntimeException):
     pass
 
 
+class InlineQueryFailedError(PolarRuntimeException):
```
Comment (gj) [inline languages/python/oso/polar/exceptions.py:48]: 😍

File: polar-core/src/vm.rs [hunk 17]
```diff
@@ -2258,36 +2258,58 @@ impl PolarVirtualMachine {
...
+        source_string
+    }
+
+    pub fn term_source_info(&self, term: &Term) -> String {
+        self.term_source(term, true)
+    }
+
+    pub fn term_source_string(&self, term: &Term) -> String {
+        self.term_source(term, false)
     }
```
Comment (gj) [inline polar-core/src/vm.rs:2292]: nit: I would personally do away with these two helpers and just call `vm.term_source(term, [true|false])` at the callsites b/c I don't love the naming of `term_source_info` and `term_source_string` but can't come up with anything better 😄
